README.md
<center> <h4> POWSC: A computational tool for power evaluation and sample size estimation in scRNA-seq </h4> </center>

-------------------
**POWSC** is a R package designed for scRNA-seq with a wild range of usage. It can play three roles: **_parameter estimator_**, **_data simulator_**, and **_power assessor_**. As the parameter estimator, POWSC accurately captures the characterized parameters (`Fig.B`) for any specific cell type from a given real expression data (`Fig.A`). As the data simulator, POWSC generates sythetic data (`Fig.C`) based on a rigorous simulation mechanism incluidng zero expression values. As the power assessor, POWSC performs comprehensive power analysis and reports the stratified targeted powers (`Fig.D`) for two forms of DE genes. A schemetic overview of the aglorithm is shown in (`Fig.E`). All the copyrights are explaned by Kenong Su <kenong.su@emory.edu> and Dr. Wu's lab <http://www.haowulab.org>.
![workflow](vignettes/workflow.png)

This tutorial introduces the basic functionalities of POWSC. Please use the <font color="blue">**vignette("POWSC")**</font> to review more detailed package vignette. It is worth noting that one might need pre-install dependent R packages such as MAST, SC2P, and SummarizedExperiment.

The corresponding paper can be found here:

---
references:
- POWSC
  title: Simulation, power evaluation and sample size recommendation for single-cell RNA-seq
  author:
  - Su
    Kenong
  URL: 'https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btaa607/5866544'
  DOI: https://doi.org/10.1093/bioinformatics/btaa607
  publisher: Bioinformatics
---

### How to get help for POWSC
Any POWSC questions should be posted to the GitHub Issue section of POWSC homepage at https://github.com/suke18/POWSC/issues.


### 1. Software Installation
```
library(devtools)
install_github("suke18/POWSC", build_vignettes = T, dependencies = T)
R CMD INSTALL POWSC_0.1.0.tar.gz # Alternatively, use this command line in the terminal.
```

### 2. Code Snippets
**(1). parameter estimation for one cell type case**
```r
library(POWSC)
data("es_mef_sce")
sce = es_mef_sce[, colData(es_mef_sce)$cellTypes == "fibro"]
est_Paras = Est2Phase(sce)
```

**(2). the first scenairo of two-group comparison**
```r
sim_size = c(100, 400, 1000) # A numeric vector
pow_rslt = runPOWSC(sim_size = sim_size, est_Paras = est_Paras,per_DE=0.05, DE_Method = "MAST", Cell_Type = "PW") # Note, using our previous developed tool SC2P is faster.
plot_POWSC(pow_rslt, Form="II", Cell_Type = "PW") # Alternatively, we can use Form="I"
summary_POWSC(pow_rslt, Form="II", Cell_Type = "PW")
```
**(3). the second scenairo of multi-group comparisons**.
The sample data can be found here: https://www.dropbox.com/s/55zdktqfqiwfs3l/GSE67835.RData?dl=0.
```r
sim_size = 1000
cell_per = c(0.2, 0.3, 0.5)
load("pathto/GSE67835.RData") #data("GSE67835")
col = colData(sce)
exprs = assays(sce)$counts
(tb = table(colData(sce)$Patients, colData(sce)$cellTypes))
# use AB_S7 patient as example and take three cell types: astrocytes hybrid and neurons
estParas_set = NULL
celltypes = c("oligodendrocytes", "hybrid", "neurons")
for (cp in celltypes){
    print(cp)
    ix = intersect(grep(cp, col$cellTypes), grep("AB_S7", col$Patients))
    tmp_mat = exprs[, ix]
    tmp_paras = Est2Phase(tmp_mat)
    estParas_set[[cp]] = tmp_paras
}
pow_rslt = runPOWSC(sim_size = sim_size, est_Paras = estParas_set,per_DE=0.05, DE_Method = "MAST",multi_Prob = cell_per, Cell_Type = "Multi")
plot_POWSC(pow_rslt, Form="I", Cell_Type = "Multi")
summary_POWSC(pow_rslt, Form="II", Cell_Type = "Multi")
```


inst/NEWS
CHANGES IN VERSION 0.99
----------------------

NEW FEATURES

  o Initial review.

CHANGES IN VERSION 0.99.0
----------------------
  o Revise required files and format the code style.




tests/testthat.R
library(testthat)
library(POWSC)

test_check("POWSC")


tests/testthat/test-all.R
# simulation of expression data
library(POWSC)
data("es_mef_sce")
sce = es_mef_sce[, colData(es_mef_sce)$cellTypes == "fibro"]
set.seed(12)
rix = sample(1:nrow(sce), 500)
sce = sce[rix, ]


# test the input data. Alternatively, it could a class of "matrix".
test_that('Check the input sce', {
    expect_that(sce, is_a('SingleCellExperiment'))
})



# test the estimated parameter
test_that("Check the output of Est2Phase", {
    est_Paras = Est2Phase(sce)
    # gene-wise
    expect_equal(length(est_Paras$mu), nrow(sce))
    expect_equal(length(est_Paras$sd), nrow(sce))
    # cell-wise
    expect_equal(length(est_Paras$sf), ncol(sce))
    expect_equal(length(est_Paras$lambda), ncol(sce))

})

# test the main function runPOWSC
test_that('Check the output of runPOWSC', {
    est_Paras = Est2Phase(sce)
    sim_size = c(100, 200) # A numeric vector
    pow_rslt = runPOWSC(sim_size = sim_size, est_Paras = est_Paras,per_DE=0.05, DE_Method = "MAST", Cell_Type = "PW")
    pow_rslt1 = pow_rslt[[1]]
    pow_rslt2 = pow_rslt[[2]]

    expect_true(all(names(pow_rslt1[[2]]) == names(pow_rslt1[[2]])))
    expect_true(all(names(pow_rslt2[[2]]) == names(pow_rslt2[[2]])))

    expect_equal(length(pow_rslt1[[1]][[1]]), length(pow_rslt2[[1]][[1]]))
    expect_equal(length(pow_rslt1[[2]][[1]]), length(pow_rslt2[[2]][[1]]))

})




R/runDE.R
#### RUN DE METHODS #####
#@param rawEset much has a pData variable called celltype indicator

# Set up global variables
model = ~ cellTypes + cngeneson # in MAST
termTotest = "cellTypes" # in SC2P


toEset = function(sce){
    y = assays(sce)$counts
    cellTypes = colData(sce)$cellTypes
    phenoData = new("AnnotatedDataFrame", data = data.frame(cellTypes = cellTypes))
    rownames(phenoData) = colnames(y)
    eset = ExpressionSet(assayData = y, phenoData = phenoData)
    return(eset)
}


##### This is the function from the https://github.com/haowulab/SC2P/blob/master/R/twoPhaseDE.R.
##### Since it is from github not CRAN or Bioconductor, I modify for POWSC usage on Bioconductor.
twoPhaseDE <- function(norm,
                       design, ## vector of covariate names
                       test.which, ## which of design to be tested?
                       low.prob=.99){
    Y <- norm$exprs
    Z <- norm$Z
    Offset <- norm$Offset

    k = colSums(Y)
    k = log2(k/median(k))
    Offset = matrix(rep(k, nrow(Y)), nrow=nrow(Y), byrow=TRUE)

    X = norm$phenoData@data[, design, drop=FALSE] ## or X can be permuted
    twoPhaseDE0(Y=Y, Z=Z, X=X, Offset=Offset,
                test.which=test.which, low.prob=low.prob)
}




#' Run DE analysis by using SC2P. Here we output two result tables corresponding to two forms of DE genes.
#'
#' @param sce is a simulated scRNA-seq dataset with two-group conditions, e.g., treatment vs control.
#' @return a list of three tables: the first table summaries the DE result for both forms of DE genes. cont table represents the result for continous case. disc table shows the result for discontinous case.
runSC2P = function(sce) {
    sce$cellTypes = factor(sce$cellTypes)
    rawEset = toEset(sce = sce); ngene = nrow(sce)
    norm = eset2Phase(rawEset)  ## eset if of class ExpressionSet or eSet
    deSC2P = twoPhaseDE(norm, design = termTotest, test.which=1)
    discPval = deSC2P$Ph1.pval; discFdr = p.adjust(discPval, method = "fdr")
    contPval = deSC2P$Ph2.pval; contFdr = p.adjust(contPval, method = "fdr")
    disLen = length(discPval)
    conLen = length(contPval)
    disc = data.frame(geneIndex = seq_len(disLen), pval = discPval, fdr = discFdr)
    cont = data.frame(geneIndex = seq_len(conLen), pval = contPval, fdr = contFdr)
    rownames(disc) = rownames(cont) = rownames(deSC2P)
    return(list(table = deSC2P, cont = cont, disc = disc))
}

#' Run DE analysis by using MAST. Here we output two result tables corresponding to two forms of DE genes.
#' These parameters include four gene-wise parameters and two cell-wise parameters.
#'
#' @param sce is a simulated scRNA-seq dataset with two-group conditions, e.g., treatment vs control.
#' @return a list of three tables: the first table summaries the DE result for both forms of DE genes. cont table represents the result for continous case. disc table shows the result for discontinous case.
runMAST = function(sce) {
    ## grab raw counts and compute log TPM
    rawEset = toEset(sce = sce)
    Y = exprs(rawEset)
    sf = colSums(Y)/1e6
    ltpm = log2(sweep(Y, 2, sf, FUN="/")+1)
    sca = FromMatrix(ltpm, pData(rawEset), fData(rawEset)) # the object is sca for MAST
    ## Start testing
    cdr2 = colSums(assay(sca)>0)
    colData(sca)$cngeneson = scale(cdr2)
    thres = thresholdSCRNACountMatrix(assay(sca)) #, nbins=200, min_per_bin=30)
    assays(sca) = list(thresh=thres$counts_threshold, tpm=assay(sca))
    fit = zlm(model, sca)
    rslt = lrTest(fit, termTotest)  # fiting zlm model
    table = rslt[,,3]
    two.des = list()
    for (i in seq_len(2)){
        pval = table[, i]
        fdr = p.adjust(pval, method = "fdr")
        result = data.frame(geneIndex=seq_len(length(pval)), pval=pval, fdr=fdr)
        res = result[complete.cases(result),]
        two.des[[i]] = res
    }
    return(list(table = table, cont = two.des[[1]], disc = two.des[[2]]))
}


#' A wrapper function for calling DE genes. This contains two methods: MAST and SC2P
#'
#' @param sce is a simulated scRNA-seq dataset with two-group conditions, e.g., treatment vs control.
#' @param DE_Method is a string chosen from "MAST" or "SC2P".
#' @return a list of three tables: the first table summaries the DE result for both forms of DE genes. cont table represents the result for continous case. disc table shows the result for discontinous case.
#' @examples
#' data("es_mef_sce")
#' sce = es_mef_sce[, colData(es_mef_sce)$cellTypes == "fibro"]
#' set.seed(123)
#' rix = sample(1:nrow(sce), 500)
#' sce = sce[rix, ]
#' estParas = Est2Phase(sce)
#' simData = Simulate2SCE(n=100, estParas1 = estParas, estParas2 = estParas)
#' sim_sce = simData$sce
#' DErslt = runDE(sim_sce)
#' @export runDE
runDE = function(sce, DE_Method = c("MAST", "SC2P")){
    DE_Method = match.arg(DE_Method)
    if (DE_Method == "MAST"){
        DE_rslt = runMAST(sce)
    }else{
        DE_rslt = runSC2P(sce)
    }
    return(DE_rslt)
}





##############################################################################
############################## other DE methods ##############################
##############################################################################
## BPSC
# runBPSC = function(rawEset) {
#     require(BPSC)
#     Y = exprs(rawEset)
#     sf = colSums(Y)/1e6
#     bp_mat = sweep(Y, 2, sf, FUN="/")+1
#     groups = as.factor(droplevels(pData(rawEset)$celltype))
#     controlIds = which(groups==levels(groups)[1])
#     design = model.matrix( ~ groups)
#     rslt = BPglm(data = bp_mat,
#                      controlIds = controlIds, design = design, coef = 2,
#                      estIntPar = FALSE)
#     table = summary(rslt)$topTable
#     pval = table[, 3]
#     pval[is.na(pval)] <- 1
#     qval  = qvalue(pval)$qvalues
#
#     ## construct results, ordered by the orignal order of genes
#     ix = sort(pval, index.return=TRUE)$ix
#     result = data.frame(geneIndex=ix, pval=pval[ix], qval=qval[ix])
#     res = result[order(result$geneIndex),]
#     return(res)
# }
#
# # SCDD
# runSCDD = function(simData){
#     require("SingleCellExperiment"); require("scDD")
#     Y = exprs(rawEset)
#     sf = colSums(Y)/1e6
#     ltpm = log2(sweep(Y, 2, sf, FUN="/")+1)
#     group = as.factor(droplevels(pData(rawEset)$celltype))
#     priorParam = list(alpha=0.01, mu0=0, s0=0.01, a0=0.01, b0=0.01)
#     sce = SingleCellExperiment(assays=list(normcounts = as.matrix(ltpm)), colData=data.frame(condition = group))
#     rslt = scDD(sce, priorParam, testZeroes=F)
#     table = results(rslt)
#     pval = table[, 7]
#     pval[is.na(pval)] <- 1
#     fdr  = qvalue::lfdr(pval)
#
#     ## construct results, ordered by the orignal order of genes
#     ix = sort(pval, index.return=TRUE)$ix
#     result = data.frame(geneIndex=ix, pval=pval[ix], fdr=fdr[ix])
#     res = result[order(result$geneIndex),]
#     return(res)
# }
#
#
# ## SCDE
# runSCDE = function(rawEset, ncores=32) {
#     require(scde)
#     Y = exprs(rawEset)
#     Y = apply(Y,2,function(x) {
#         storage.mode(x) <- 'integer'; x})
#     colnames(Y) = 1:ncol(Y)
#     cd = clean.counts(Y)
#     sg = (pData(rawEset)$celltype)[as.integer(colnames(cd))]
#     o.ifm = scde.error.models(counts=cd, groups=sg, n.cores=ncores,
#                                threshold.segmentation = TRUE,
#                                save.crossfit.plots = FALSE,
#                                save.model.plots = FALSE, verbose = 1)
#
#     valid.cells = o.ifm$corr.a > 0
#     table(valid.cells)
#     o.ifm = o.ifm[valid.cells, ]
#     o.prior = scde.expression.prior(models = o.ifm,
#                                      counts = cd, length.out = 400,
#                                      show.plot = FALSE)
#     groups = sg[valid.cells]
#     de.scde = scde.expression.difference(o.ifm, cd, o.prior,
#                                           groups  =  groups,
#                                           n.randomizations  =  100,
#                                           n.cores  =  ncores,
#                                           verbose  =  0)
#     return(de.scde)
# }
#
#
# makeROC = function(pval, trueDE) {
#     pred = prediction(-log10(pval), trueDE)
#     auc = performance(pred,"auc")
#     auc = unlist(slot(auc, "y.values"))
#     perf = performance(pred,"tpr","fpr")
#     return(list(perf = perf, auc = auc))
# }


### Main function #### should rewrite this function
# runDE = function(simData){
#     rawEset = simData$eset
#     rslt = runMAST(rawEset)
#     pvals = rslt$pval
#     trueDE = sign(match(rownames(rslt), simData$DEGs, nomatch = 0))
#     roc_rslt = makeROC(rslt$pval, trueDE)
#     return(list(simData = simData,
#                 auc = roc_rslt$auc, perf = roc_rslt$perf))
# }




R/estPara.R
#' Estimate characterized parameters for a given scRNA-seq data (SingleCellExperiment object or a count matrix).
#'
#' These parameters include four gene-wise parameters and two cell-wise parameters.
#'
#' @param sce SingleCellExperiment object with assays(sce)[[1]] is the count matrix or input directly
#' @param low.prob lower bound probability for phase I
#' @return a list of needed estimated parameters
#' @examples
#' data("es_mef_sce")
#' sce = es_mef_sce[, colData(es_mef_sce)$cellTypes == "fibro"]
#' set.seed(123)
#' rix = sample(1:nrow(sce), 500)
#' sce = sce[rix, ]
#' estParas = Est2Phase(sce)
#' @export Est2Phase
Est2Phase = function(sce, low.prob=0.99){
    # check the input whether it is an SingleCellExperiment Object or a matrix
    if (is(sce, "SingleCellExperiment")){
        Y = round(assays(sce)[[1]])
    }else{Y = round(sce)}
    ## ## initial estimate of prob(X\in bg)
    Cell0=colMeans(Y==0) # each cell has this percentage 0
    par1= apply(Y,2,function(yy) {
        yyy=yy[yy<=15]
        ### consider (0s, 3, 4)
        if (length(table(yyy)) <= 2){
            # print(which(colSums(Y==yy) == nrow(Y)))
            yyy = yy[yy<=20]
        }
        RobustPoi0(yyy)
        })
    ncell = ncol(sce)
    pi0.hat=Cell0/(par1[1,]+(1-par1[1,])*dpois(0,par1[2,]))
    if (any((pi0.hat > 1))) {warning("Zero proportion is greater than estimation.")}
    pi0.hat <- pmin(pi0.hat, 1)
    prob0=pi0.hat*par1[1,]+ pi0.hat*(1-par1[1,])*dpois(0,par1[2,]) ## ZIP prob at 0
    ## First round
    ## get the 1-low.prob quantile of ZIP
    x0=qpois(pmax(1-(1-low.prob)/(1-par1[1,]),0),par1[2,])
    Z= sweep(Y,2,x0)>0 # indicate if a gene is > bg
    L=colSums(Y*Z)/1e6 # so far it is like simple total..

    mu.g1=log2(rowSums(Z*Y)/rowSums(sweep(Z,2,L,FUN="*")))
    mu.g1[is.na(mu.g1)]=0 ## if allZ is 0, it gets NA,
    ### but we should shrink mu.g1 as well since some mu.g1 is estimated by only a few observations
    ## leave it here for now.
    n.g1=rowSums(Z)
    y1=log2(sweep(Y,2,L,FUN="/")+1) #like TPM**
    s.g1=sqrt(rowSums(Z*sweep(y1,1,mu.g1)^2)/(n.g1-1)) ## TPM type of SD
    mu.g2 = shrink.mu(mu.g1,s.g1,n.g1)
    ## get sd.g
    res.g1=log2(sweep(Y,2,L,FUN="/")+1)-mu.g1
    ## mad of those res.g1 that are associated with Z==1
    tmp=array(0,dim=c(dim(res.g1),2))
    tmp[,,1]=res.g1;tmp[,,2]=Z
    sd.g1=apply(tmp,1,function(xx) my.mad(xx[xx[,2]==1,1]))
    sd.g1[is.na(sd.g1)]=0## if all bg, there's no info about fg sd
    ## add a shrinkage for sd.g1
    sd.prior=squeezeVar(sd.g1^2,n.g1-1)
    sd.g2=sqrt(sd.prior$var.post)
    #####  gene specific bg. Z_gi
    den.fg = den.bg = NA*Y
    for(i in seq_len(ncell)){
        den.bg[,i]=dZinf.pois(Y[,i], par1[1,i], par1[2,i])
        den.fg[,i]=dLNP2(x=Y[,i], mu=mu.g1, sigma=sd.g2, l=L[i])
    }
    Z.fg=sweep(den.fg,2,1-pi0.hat,FUN="*")
    Z.bg=sweep(den.bg,2,pi0.hat,FUN="*")
    post.Z=Z.fg/(Z.fg+Z.bg)
    post.Z[is.na(post.Z)] <- 1

    ### if I shrink mu.g
    den.fg2 = NA*Y
    for (i in seq_len(ncell)){
        den.fg2[,i]= dLNP2(x=Y[,i], mu=mu.g2, sigma=sd.g2, l=L[i])
    }
    Z.fg2=sweep(den.fg2,2,1-pi0.hat,FUN="*")
    post.Z2=Z.fg2/(Z.fg2+Z.bg)
    post.Z2[is.na(post.Z2)] <- 1

    pi.g = rowMeans(post.Z2)
    est_rslt = list(exprs = Y, pi.g = pi.g, p0 = par1[1,], lambda = par1[2,],
                    mu = mu.g2, sd = sd.g2, sf = L)
    return(est_rslt)
}





R/runPOWSC.R
#' Estimate characterized parameters for a given scRNA-seq data (SingleCellExperiment object or a count matrix).
#'
#' These parameters include four gene-wise parameters and two cell-wise parameters.
#'
#' @param sim_size a list of numbers
#' @param per_DE the percentage of the DE genes.
#' @param est_Paras the template parameter estimated from one cell type.
#' @param DE_Method is a string chosen from "MAST" or "SC2P".
#' @param Cell_Type is a string corresponding to the 1st scenario: same cell type comparison, and 2nd scenario: multiple cell types.
#' @param multi_Prob is the mixture cell proportions which sum up to 1. If not summing up to 1, then the package will internally do the normalization procedure.
#' @param alpha is the cutoff for the fdr which can be modified
#' @param disc_delta or the zero ratio change is the cutoff (=0.1) used to determined the high DE genes for Form II.
#' @param cont_delta or the lfc is the cutoff (=0.5) used to determined the high DE genes for Form II.
#' @return POWSC object
#' @examples
#' data("es_mef_sce")
#' sce = es_mef_sce[, colData(es_mef_sce)$cellTypes == "fibro"]
#' set.seed(12)
#' rix = sample(1:nrow(sce), 500)
#' sce = sce[rix, ]
#' est_Paras = Est2Phase(sce)
#' sim_size = c(100, 200) # A numeric vector
#' pow_rslt = runPOWSC(sim_size = sim_size, est_Paras = est_Paras,per_DE=0.05, DE_Method = "MAST", Cell_Type = "PW") # Note, using our previous developed tool SC2P is faster.
#' @export
runPOWSC = function(sim_size = c(50, 100, 200, 800, 1000), per_DE = 0.05, est_Paras, DE_Method = c("MAST", "SC2P"),
                    Cell_Type = c("PW", "Multi"), multi_Prob = NULL,
                    alpha = 0.1, disc_delta = 0.1, cont_delta = 0.5){
    DE_Method = match.arg(DE_Method)
    Cell_Type = match.arg(Cell_Type)
    pow_rslt = NULL
    # Pair-wise comparison
    if (Cell_Type == "PW"){
        for (tmp_size in sim_size){
            sim_Data = Simulate2SCE(n = tmp_size, estParas1 = est_Paras, estParas2 = est_Paras)
            DE_rslt = runDE(sim_Data$sce, DE_Method = DE_Method)
            pow1_rslt = Power_Disc(DErslt = DE_rslt, simData = sim_Data, alpha = alpha, delta = disc_delta)
            pow2_rslt = Power_Cont(DErslt = DE_rslt, simData = sim_Data, alpha = alpha, delta = cont_delta)
            tmp_rslt = list(pow1 = pow1_rslt, pow2 = pow2_rslt)
            pow_rslt[[toString(tmp_size)]] = tmp_rslt
        }
    }else{ # Multiple Cell Types
        if (missing(multi_Prob)){
            stop("Please Estimate the Cell Proportions by SC3 or Seurat")
        }
        for (tmp_size in sim_size){
            pow1_rslt = pow2_rslt = NULL
            sim_Data = SimulateMultiSCEs(n = tmp_size, estParas_set = est_Paras, multiProb = multi_Prob)
            for (comp in names(sim_Data)){
                tmp_DE = runDE(sim_Data[[comp]]$sce, DE_Method = DE_Method)
                pow1_rslt[[comp]] = Power_Disc(DErslt = tmp_DE, simData = sim_Data[[comp]], alpha = alpha, delta = disc_delta)
                pow2_rslt[[comp]] = Power_Cont(DErslt = tmp_DE, simData = sim_Data[[comp]], alpha = alpha, delta = cont_delta)
            }
            tmp_rslt = list(pow1 = pow1_rslt, pow2 = pow2_rslt)
            pow_rslt[[toString(tmp_size)]] = tmp_rslt
        }
    }
    class(pow_rslt) = "POWSC"
    return(pow_rslt)
}






#' plot the result use visualization.
#'
#' @param POWSCobj a POWSC object from \code{runPOWSC}
#' @param Form choose from "I" or "II".
#' @param Cell_Type choose from "PW" or "Multi"
#' @return for multiple comparison cases, return the pheatmap; for the pairwise comparison, return the ggplot object.
#' @examples
#' data("es_mef_sce")
#' sce = es_mef_sce[, colData(es_mef_sce)$cellTypes == "fibro"]
#' set.seed(12)
#' rix = sample(1:nrow(sce), 500)
#' sce = sce[rix, ]
#' est_Paras = Est2Phase(sce)
#' sim_size = c(100, 200) # A numeric vector
#' pow_rslt = runPOWSC(sim_size = sim_size, est_Paras = est_Paras,per_DE=0.05, DE_Method = "MAST", Cell_Type = "PW") # Note, using our previous developed tool SC2P is faster.
#' plot_POWSC(pow_rslt, Form="II", Cell_Type = "PW") # Alternatively, we can use Form="I")
#' @export
plot_POWSC = function(POWSCobj, Form = c("I", "II"), Cell_Type = c("PW", "Multi")){
    Cell_Type = match.arg(Cell_Type)
    Form = match.arg(Form)
    if (Cell_Type == "Multi"){
        colors = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)
        sim_size = names(POWSCobj)
        for (tmp_size in sim_size){
            tmp_pow_rslt = POWSCobj[[tmp_size]]
            pow1_mat = do.call(rbind, lapply(tmp_pow_rslt$pow1, function(x) x$power))
            pow2_mat = do.call(rbind, lapply(tmp_pow_rslt$pow2, function(x) x$power))
            colnames(pow1_mat) = names(POWSCobj[[1]][[1]][[1]][[1]])
            colnames(pow2_mat) = names(POWSCobj[[1]][[2]][[1]][[1]])
            if (Form == "I"){
                pow = pow1_mat
            }else{
                pow = pow2_mat
            }

            pheatmap(pow,display_numbers = TRUE, color=colors, show_rownames = FALSE,
                     show_colnames = TRUE, fontsize = 20, fontsize_col = 20,
                     cellwidth = 40, cellheight = 40, legend = FALSE,
                     border_color = "grey96", na_col = "grey",
                     cluster_rows = FALSE, cluster_cols = FALSE,
                     breaks = seq(0, 1, 0.01),
                     main = paste0("Total Cell Number = ", tmp_size))
        }
    }
    else{
        pow1_mat = do.call(rbind, lapply(POWSCobj, function(x) x$pow1$power))
        pow2_mat = do.call(rbind, lapply(POWSCobj, function(x) x$pow2$power))
        nrep = rownames(pow1_mat)
        nm1 = names(POWSCobj[[1]][[1]][[1]])
        nm2 = names(POWSCobj[[1]][[2]][[1]])
        pow1 = data.frame(Strata = rep(nm1, each = nrow(pow1_mat)), Power = as.vector(pow1_mat), Reps = rep(nrep, ncol(pow1_mat)), stringsAsFactors = FALSE)
        pow2 = data.frame(Strata = rep(nm2, each = nrow(pow2_mat)), Power = as.vector(pow2_mat), Reps = rep(nrep, ncol(pow2_mat)), stringsAsFactors = FALSE)
        pow1$Strata = factor(pow1$Strata, levels = nm1); pow2$Strata = factor(pow2$Strata, levels = nm2)
        pow1$Reps = factor(pow1$Reps, levels = unique(pow1$Reps)); pow2$Reps = factor(pow2$Reps, levels = unique(pow2$Reps))
        if (Form == "I"){
            pow = pow1
            tit = ggtitle("Form I DE genes")
            tmpxlab = xlab("strata of zero ratios")
        }else{
            pow = pow2
            tit = ggtitle("Form II DE genes")
            tmpxlab = xlab("strata of average reads")
        }
        breaks = round(seq(0, 1, length = 6), 1)
        ggplot(pow, aes(x=Strata, y=Power, group=Reps, color=Reps)) +
            geom_line(aes(color=Reps, linetype = Reps), size = 0.6)+
            geom_point(aes(color=Reps, shape = Reps), size = 1.5) +
            scale_colour_brewer(palette = "Set1") +
            tmpxlab + tit + theme_classic()  +
            scale_y_continuous(breaks=breaks, limits = c(0, 1+0.1), labels=format(breaks, nsmall = 1))+
            theme(legend.position = c(0.5, 0.95), legend.direction = "horizontal") +
            theme(plot.title = element_text(size=16, face="bold"),
                  axis.text.x=element_text(colour="black", size = 14),
                  axis.text.y=element_text(colour="black", size = 14),
                  legend.title = element_text(size = 15),
                  legend.text = element_text(size = 15),
                  axis.title.y = element_text(size = 15))
    }
}





#' summary of the result
#'
#' @param POWSCobj a POWSC object from \code{runPOWSC}
#' @param Form choose from "I" or "II".
#' @param Cell_Type choose from "PW" or "Multi".
#' @return return the summary of the power including stratified, marginal, and overall power.
#' @examples
#' data("es_mef_sce")
#' sce = es_mef_sce[, colData(es_mef_sce)$cellTypes == "fibro"]
#' set.seed(12)
#' rix = sample(1:nrow(sce), 500)
#' sce = sce[rix, ]
#' est_Paras = Est2Phase(sce)
#' sim_size = c(100, 200) # A numeric vector
#' pow_rslt = runPOWSC(sim_size = sim_size, est_Paras = est_Paras,per_DE=0.05, DE_Method = "MAST", Cell_Type = "PW") # Note, using our previous developed tool SC2P is faster.
#' summary_POWSC(pow_rslt, Form="II", Cell_Type = "PW")
#' @export
summary_POWSC = function(POWSCobj, Form = c("I", "II"), Cell_Type = c("PW", "Multi")){
    Cell_Type = match.arg(Cell_Type)
    Form = match.arg(Form)
    if (Cell_Type == "Multi"){
        pow_tab = NULL
        sim_size = names(POWSCobj)
        for (tmp_size in sim_size){
            tmp_pow_rslt = POWSCobj[[tmp_size]]
            pow1_mat = do.call(rbind, lapply(tmp_pow_rslt$pow1, function(x) x$power))
            pow2_mat = do.call(rbind, lapply(tmp_pow_rslt$pow2, function(x) x$power))
            colnames(pow1_mat) = names(POWSCobj[[1]][[1]][[1]][[1]])
            colnames(pow2_mat) = names(POWSCobj[[1]][[2]][[1]][[1]])
            if (Form == "I"){
                pow = pow1_mat
            }else{
                pow = pow2_mat
            }
            pow_tab[[tmp_size]] = pow
        }
        return(pow_tab)
    }else{
        pow1_mat = round(do.call(rbind, lapply(POWSCobj, function(x) x$pow1$power)), 4)
        pow2_mat = round(do.call(rbind, lapply(POWSCobj, function(x) x$pow2$power)), 4)
        nm1 = names(POWSCobj[[1]][[1]][[1]])
        nm2 = names(POWSCobj[[1]][[2]][[1]])
        colnames(pow1_mat) = nm1
        colnames(pow2_mat) = nm2
        if (Form == "I"){
            tab = pow1_mat
        }else{
            tab = pow2_mat
        }
        return(tab)
    }
}

